### DISCUSSION TITLE
how to get the shell auto-complete files

### DISCUSSION THREAD

**Discussion by topazus** [ClementTsang/bottom#792] [Q&A] [answered]
Is there a way to get the shell auto-complete files from github repository?

**Comment by ClementTsang** [comment, marked as answer]
They are automatically included in the stable and nightly releases (e.g. https://github.com/ClementTsang/bottom/releases/download/0.6.8/completion.tar.gz)

**Reply by topazus** [reply]
Thanks for the reply. It would be much better if you could include these shell completion files in github repo.

**Reply by ClementTsang** [reply]
I probably won't do that as these files are auto-generated on build, sorry.

**Reply by topazus** [reply]
> I probably won't do that as these files are auto-generated on build, sorry.

if i want to get in building bottom from this git source repo, how can I generate these completion files by which command?

**Reply by ClementTsang** [reply]
They're generated by the [`build.rs`](https://github.com/ClementTsang/bottom/blob/master/build.rs) file when the `BTM_GENERATE` env var is set to anything. So off the top of my head, you would want to do something like:

```
BTM_GENERATE=true cargo build
```
which generates manpage + completion files.

You can see how this is done in CI [here](https://github.com/ClementTsang/bottom/blob/master/.github/workflows/build_releases.yml#L22).

There's a tiny bit of a hack needed in that if you generate once, then want to generate again, you may need to set the env var to something different to force the `build.rs` script to retrigger (or update either the `clap.rs` or `build.rs` files) - this isn't super great but it works for the use cases I imagined most people would need to generate completions for.

---

This actually reminds me that I should really add/update some documentation on this, I guess I forgot since this functionality was only ever intended for use in CI. Will do that soon.

**Reply by ClementTsang** [reply]
Updated documentation on how to build manpages/completion found [here](https://clementtsang.github.io/bottom/nightly/contribution/packaging-and-distribution/#manpage-and-completion-generation).
